### PR TITLE
Feature/filter terms

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

File: core/search/phrase_filter.hpp [hunk 1]
```diff
@@ -41,13 +41,10 @@ class by_phrase;
...
  private:
-  using phrase_part = std::variant<
-    by_term_options,
-    by_prefix_options,
-    by_wildcard_options,
-    by_edit_distance_filter_options,
-    by_terms_options,
-    by_range_options>;
+  using phrase_part =
+      std::variant<by_term_options, by_prefix_options, by_wildcard_options,
```
Comment (MBkkt) [inline core/search/phrase_filter.hpp:45]: maybe manual format them

File: core/search/phrase_filter.hpp [hunk 2]
```diff
@@ -58,7 +55,8 @@ class by_phrase_options {
   // Returns reference to the inserted phrase part.
   template<typename PhrasePart>
   PhrasePart& insert(size_t pos) {
-    is_simple_term_only_ &= std::is_same<PhrasePart, by_term_options>::value; // constexpr
+    is_simple_term_only_ &=
+        std::is_same<PhrasePart, by_term_options>::value;  // constexpr
```
Comment (MBkkt) [inline core/search/phrase_filter.hpp:59]: remove useless comment
compiler should optimize this

File: core/search/phrase_filter.hpp [hunk 3]
```diff
@@ -67,7 +65,8 @@ class by_phrase_options {
   // Returns reference to the inserted phrase part
   template<typename PhrasePart>
   PhrasePart& insert(PhrasePart&& t, size_t pos) {
-    is_simple_term_only_ &= std::is_same<PhrasePart, by_term_options>::value; // constexpr
+    is_simple_term_only_ &=
+        std::is_same<PhrasePart, by_term_options>::value;  // constexpr
```
Comment (MBkkt) [inline core/search/phrase_filter.hpp:69]: remove useless comment
compiler should optimize this

File: core/search/phrase_query.hpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  doc_iterator::ptr execute(const ExecutionContext& ctx) const override;
+
+  doc_iterator::ptr ExecuteWithOffsets(const irs::sub_reader& segment) const;
+
+  void visit(const sub_reader& segment, PreparedStateVisitor& visitor,
+             score_t boost) const override {
+    if (auto state = states_.find(segment); state) {
+      visitor.Visit(*this, *state, boost);
+    }
+  }
```
Comment (MBkkt) [inline core/search/phrase_query.hpp:74]: duplication

File: core/search/states/multiterm_state.hpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+struct MultiTermState {
+  struct ScoredTermState {
+    ScoredTermState(seek_cookie::ptr&& cookie, uint32_t stat_offset,
+                    score_t boost = kNoBoost) noexcept
+        : cookie{std::move(cookie)}, stat_offset{stat_offset}, boost{boost} {}
+
+    seek_cookie::ptr cookie;
+    uint32_t stat_offset{};
+    float_t boost{kNoBoost};
```
Comment (MBkkt) [inline core/search/states/multiterm_state.hpp:41]: is it necessary?

File: core/search/phrase_query.cpp [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+    ++position;
+  }
+
+  if (ord.empty()) {
+    return memory::make_managed<FixedPhraseIterator<true, false>>(
+        std::move(itrs), std::move(positions));
+  }
+
+  return memory::make_managed<::FixedPhraseIterator<false, true>>(
```
Comment (MBkkt) [inline core/search/phrase_query.cpp:110]:
```suggestion
  return memory::make_managed<FixedPhraseIterator<false, true>>(
```

File: core/search/phrase_query.cpp [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+    conjunction<doc_iterator::ptr, NoopAggregator>,
+    VariadicPhraseFrequency<Adapter, VolatileBoost, OneShot, HasFreq>>;
+
+}  // namespace
+
+namespace iresearch {
+
+doc_iterator::ptr FixedPhraseQuery::execute(const ExecutionContext& ctx) const {
+  auto& rdr = ctx.segment;
+  auto& ord = ctx.scorers;
```
Comment (MBkkt) [inline core/search/phrase_query.cpp:53]: nit
move to first usage

**Review by MBkkt — APPROVED** [review]

**Review by Dronplane — APPROVED** [review]
LGTM  only NGramPosition::next needs some attention

File: core/search/nested_filter.cpp [hunk 8]
```diff
@@ -594,6 +594,11 @@ class ByNesterQuery final : public filter::prepared {
   using filter::prepared::execute;
   doc_iterator::ptr execute(const ExecutionContext& ctx) const override;
 
+  void visit(const sub_reader& segment, PreparedStateVisitor& visitor,
+             score_t boost) const override {
+    child_->visit(segment, visitor, this->boost() * boost);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/nested_filter.cpp:599]: ensured by ctor

File: core/search/ngram_similarity_query.cpp [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+      return &offset_;
+    }
+
+    return nullptr;
+  }
+
+  bool next() final {
+    if (begin_ == std::end(offsets_)) {
+      offsets_.clear();
+      begin_ = std::end(offsets_);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/ngram_similarity_query.cpp:117]: fixed

File: core/search/ngram_similarity_query.cpp [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+    }
+  } else {
+    return 0;
+  }
+}
+
+struct SearchState {
+  template<typename T>
+  SearchState(uint32_t p, const T& attrs)
+      : parent{nullptr},
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/ngram_similarity_query.cpp:74]: done

File: core/search/phrase_filter.hpp (hunk 3 above)
Comment (gnusi) [reply core/search/phrase_filter.hpp:69]: done

File: core/search/phrase_filter.hpp (hunk 1 above)
Comment (gnusi) [reply core/search/phrase_filter.hpp:45]: let's do it after the global reformat

File: core/search/phrase_query.cpp (hunk 6 above)
Comment (gnusi) [reply core/search/phrase_query.cpp:110]: done

**Pull request merged**